ambry-router/src/main/java/com/github/ambry/router/ParanoidDurabilityOperationTracker.java

```diff
@@ -0,0 +1,314 @@
+      NonBlockingRouterMetrics routerMetrics) {
+    super(routerConfig, RouterOperation.PutOperation, partitionId, originatingDcName, true, routerMetrics);
+
+    localReplicaParallelism = routerConfig.routerPutLocalRequestParallelism;
+    remoteReplicaParallelism = routerConfig.routerPutRemoteRequestParallelism;
+    localReplicaSuccessTarget = routerConfig.routerPutLocalSuccessTarget;
+    remoteReplicaSuccessTarget = routerConfig.routerPutRemoteSuccessTarget;
+
+    List<ReplicaId> allEligibleReplicas = getEligibleReplicas(datacenterName, EnumSet.of(ReplicaState.STANDBY, ReplicaState.LEADER));
+    addReplicasToPool(allEligibleReplicas);
+    this.otIterator = new ParanoidDurabilityTrackerIterator();
+  }
+
+  private void addReplicasToPool(List<? extends ReplicaId> replicas) {
+    Collections.shuffle(replicas);
+
+    for (ReplicaId replicaId : replicas) {
+      if (!replicaId.isDown()) {
+          addToBeginningOfPool(replicaId);
+      } else {
+        addToEndOfPool(replicaId); // As a last ditch attempt, we may still try to write to a down replica.
+      }
+    }
+
+    if(!enoughReplicas()) {
+      throw new IllegalArgumentException(generateErrorMessage(partitionId));
+    }
+    initializeTracking();
+  }
+
+  private void initializeTracking() {
+    replicaSuccessCountByDc = new HashMap<String, Integer>();
+    replicaInPoolOrFlightCountByDc = new HashMap<String, Integer>();
+    remoteDatacenters = new ArrayList<String>();
+
+    replicaPoolByDc.forEach((dcName, v) -> {
+      failedCountByDc.put(dcName, 0);
+      replicaSuccessCountByDc.put(dcName, 0);
+      replicaInPoolOrFlightCountByDc.put(dcName, 0);
+      disabledCountByDc.put(dcName, 0);
+      totalReplicaCountByDc.put(dcName, v.size());
+      allDatacenters.add(dcName);
+      if(!dcName.equals(datacenterName)) {
+        remoteDatacenters.add(dcName);
+      }
+    });
+  }
+
+  private String generateErrorMessage(PartitionId partitionId) {
+    StringBuilder errMsg = new StringBuilder("Partition " + partitionId + " has too few replicas to satisfy paranoid durability requirements. ");
+    replicaPoolByDc.forEach((dcName, v) -> {
+      errMsg.append("Datacenter: ").append(dcName).append(" has ").append(v.size()).append(" replicas in LEADER or STANDBY. ");
+    });
+    return errMsg.toString();
+  }
+
+  private boolean enoughReplicas() {
+    for (String currentDc : replicaPoolByDc.keySet()) {
+      if(currentDc.equals(datacenterName)) {
+        if (replicaPoolByDc.get(currentDc).size() < localReplicaSuccessTarget) {
+          logger.error("Not enough replicas in local data center for partition " + partitionId +
+              " to satisfy paranoid durability requirements " +
+              "(wanted " + localReplicaSuccessTarget + ", but found " + replicaPoolByDc.get(currentDc).size() + ")");
+          return false;
+        }
+      } else {
+        if (replicaPoolByDc.get(currentDc).size() < remoteReplicaSuccessTarget) {
```

Comment:
The success target should not be per dc for remote dcs.
If we have two remote dcs, and the success target is 1, as long as we have 1 replica available from either DC, it would be good. But I guess this is fine. The remote success target should be a low number for both dcs.

Reply:
Yeah, as discussed below I changed the tracking of replicas from per-DC tracking to just tracking local and remote, so this should be fixed now.